Add WebAssembly support

I am using `good_lp` in a web project and I noticed [this document](https://docs.rs/crate/good_lp/latest) says highs doesn't support wasm targets, but upstream`ERGO-Code/HiGHS` seems to have wasm support [build_webdemo.sh](https://github.com/ERGO-Code/HiGHS/blob/master/build_webdemo.sh)

### Reproduction

current behavior when attempting to compile

```bash
xks@xks:~/repos/end2$ cargo build --target wasm32-unknown-unknown
   Compiling highs-sys v1.12.1
error: failed to run custom build command for `highs-sys v1.12.1`

Caused by:
  process didn't exit successfully: `/home/xks/repos/end2/target/debug/build/highs-sys-abc5c91127052a9d/build-script-build` (exit status: 101)
  --- stdout
  CMAKE_TOOLCHAIN_FILE_wasm32-unknown-unknown = None
  CMAKE_TOOLCHAIN_FILE_wasm32_unknown_unknown = None
  TARGET_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_wasm32-unknown-unknown = None
  CMAKE_GENERATOR_wasm32_unknown_unknown = None
  TARGET_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_wasm32-unknown-unknown = None
  CMAKE_PREFIX_PATH_wasm32_unknown_unknown = None
  TARGET_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_wasm32-unknown-unknown = None
  CMAKE_wasm32_unknown_unknown = None
  TARGET_CMAKE = None
  CMAKE = None
  -- The CXX compiler identification is Clang 18.1.3
  -- The C compiler identification is Clang 18.1.3
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - failed
  -- Check for working CXX compiler: /usr/bin/clang++
  -- Check for working CXX compiler: /usr/bin/clang++ - broken
  -- Configuring incomplete, errors occurred!

  --- stderr
  running: cd "/home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out/build" && CMAKE_PREFIX_PATH="" LC_ALL="C" "cmake" "/home/xks/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/highs-sys-1.12.1/HiGHS" "-B" "/home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out/build" "-DFAST_BUILD=ON" "-DBUILD_SHARED_LIBS=OFF" "-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDLL" "-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=FALSE" "-DZLIB=OFF" "-DCMAKE_SYSTEM_NAME=unknown" "-DCMAKE_SYSTEM_PROCESSOR=wasm32" "-DCMAKE_INSTALL_PREFIX=/home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fno-exceptions --target=wasm32-unknown-unknown -w" "-DCMAKE_C_COMPILER=/usr/bin/clang" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fno-exceptions --target=wasm32-unknown-unknown -w" "-DCMAKE_CXX_COMPILER=/usr/bin/clang++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fno-exceptions --target=wasm32-unknown-unknown -w" "-DCMAKE_ASM_COMPILER=/usr/bin/clang" "-DCMAKE_BUILD_TYPE=Release"
  System is unknown to cmake, create:
  Platform/unknown to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
  Your CMakeCache.txt file was copied to CopyOfCMakeCache.txt. Please post that file on discourse.cmake.org.
  System is unknown to cmake, create:
  Platform/unknown to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
  System is unknown to cmake, create:
  Platform/unknown to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
  CMake Error at /usr/share/cmake-3.28/Modules/CMakeTestCXXCompiler.cmake:60 (message):
    The C++ compiler

      "/usr/bin/clang++"

    is not able to compile a simple test program.

    It fails with the following output:

      Change Dir: '/home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out/build/CMakeFiles/CMakeScratch/TryCompile-OGje0t'
      
      Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_47ac3/fast
      /usr/bin/gmake  -f CMakeFiles/cmTC_47ac3.dir/build.make CMakeFiles/cmTC_47ac3.dir/build
      gmake[1]: Entering directory '/home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out/build/CMakeFiles/CMakeScratch/TryCompile-OGje0t'
      Building CXX object CMakeFiles/cmTC_47ac3.dir/testCXXCompiler.cxx.obj
      /usr/bin/clang++   -ffunction-sections -fdata-sections -fno-exceptions --target=wasm32-unknown-unknown -w  -MD -MT CMakeFiles/cmTC_47ac3.dir/testCXXCompiler.cxx.obj -MF CMakeFiles/cmTC_47ac3.dir/testCXXCompiler.cxx.obj.d -o CMakeFiles/cmTC_47ac3.dir/testCXXCompiler.cxx.obj -c /home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out/build/CMakeFiles/CMakeScratch/TryCompile-OGje0t/testCXXCompiler.cxx
      Linking CXX executable cmTC_47ac3
      /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_47ac3.dir/link.txt --verbose=1
      /usr/bin/clang++  -ffunction-sections -fdata-sections -fno-exceptions --target=wasm32-unknown-unknown -w  CMakeFiles/cmTC_47ac3.dir/testCXXCompiler.cxx.obj -o cmTC_47ac3 
      clang++: error: unable to execute command: Executable "wasm-ld-18" doesn't exist!
      clang++: error: linker command failed with exit code 1 (use -v to see invocation)
      gmake[1]: *** [CMakeFiles/cmTC_47ac3.dir/build.make:100: cmTC_47ac3] Error 1
      gmake[1]: Leaving directory '/home/xks/repos/end2/target/wasm32-unknown-unknown/debug/build/highs-sys-5a38edd1944b2783/out/build/CMakeFiles/CMakeScratch/TryCompile-OGje0t'
      gmake: *** [Makefile:127: cmTC_47ac3/fast] Error 2
      
      

    

    CMake will not be able to correctly generate this project.
  Call Stack (most recent call first):
    CMakeLists.txt:23 (project)



  thread 'main' (1688573) panicked at /home/xks/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cmake-0.1.57/src/lib.rs:1132:5:

  command did not execute successfully, got: exit status: 1

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -1,6 +1,7 @@
 name: Rust
 
 on:
+  workflow_dispatch:
   push:
     branches: [ master ]
   pull_request:

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -1,7 +1,6 @@
 name: Rust
 
 on:
-  workflow_dispatch:
   push:
     branches: [ master ]
   pull_request:

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -65,5 +65,4 @@ jobs:
     - name: Build wasm32-unknown-emscripten
       run: |
         rustup target add wasm32-unknown-emscripten
-        cargo clean
-        cargo build --release --target wasm32-unknown-emscripten
+        cargo build --target wasm32-unknown-emscripten

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -51,7 +51,7 @@ jobs:
   wasm-emscripten:
     runs-on: ubuntu-latest
     env:
-      EMSDK_VERSION: 4.0.23
+      EMSDK_VERSION: 5.0.0
       EMSDK_CACHE_FOLDER: emsdk-cache
     steps:
     - uses: actions/checkout@v4

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -58,7 +58,7 @@ jobs:
       with:
         submodules: recursive
     - name: Setup emsdk
-      uses: mymindstorm/setup-emsdk@v14
+      uses: pyodide/setup-emsdk@v15
       with:
         version: ${{ env.EMSDK_VERSION }}
         actions-cache-folder: ${{ env.EMSDK_CACHE_FOLDER }}

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -50,18 +50,20 @@ jobs:
 
   wasm-emscripten:
     runs-on: ubuntu-latest
+    env:
+      EMSDK_VERSION: 4.0.23
+      EMSDK_CACHE_FOLDER: emsdk-cache
     steps:
     - uses: actions/checkout@v4
       with:
         submodules: recursive
-    - name: Install emsdk
-      run: |
-        git clone https://github.com/emscripten-core/emsdk.git
-        ./emsdk/emsdk install latest
-        ./emsdk/emsdk activate latest
+    - name: Setup emsdk
+      uses: mymindstorm/setup-emsdk@v14
+      with:
+        version: ${{ env.EMSDK_VERSION }}
+        actions-cache-folder: ${{ env.EMSDK_CACHE_FOLDER }}
     - name: Build wasm32-unknown-emscripten
       run: |
-        EMSDK_QUIET=1 source ./emsdk/emsdk_env.sh
         rustup target add wasm32-unknown-emscripten
         cargo clean
         cargo build --release --target wasm32-unknown-emscripten

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -47,3 +47,21 @@ jobs:
         cd ..
         cargo clean
         cargo build --no-default-features --features "discover"
+
+  wasm-emscripten:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        submodules: recursive
+    - name: Install emsdk
+      run: |
+        git clone https://github.com/emscripten-core/emsdk.git
+        ./emsdk/emsdk install latest
+        ./emsdk/emsdk activate latest
+    - name: Build wasm32-unknown-emscripten
+      run: |
+        EMSDK_QUIET=1 source ./emsdk/emsdk_env.sh
+        rustup target add wasm32-unknown-emscripten
+        cargo clean
+        cargo build --release --target wasm32-unknown-emscripten

--- a/build.rs
+++ b/build.rs
@@ -41,17 +41,19 @@ fn generate_bindings<'a>(
         .fold(bindgen::Builder::default(), |builder, path| {
             builder.clang_arg(format!("-I{}", path.to_string_lossy()))
         });
-    if env::var("TARGET")
-        .map(|target| target.contains("emscripten"))
-        .unwrap_or(false)
-    {
+
+    if env::var("TARGET").unwrap().contains("emscripten") {
         println!("cargo:rerun-if-env-changed=EMSDK");
-        if let Ok(emsdk) = env::var("EMSDK") {
-            let sysroot = Path::new(&emsdk).join("upstream/emscripten/cache/sysroot");
-            if sysroot.is_dir() {
-                builder = builder.clang_arg(format!("--sysroot={}", sysroot.display()));
-            }
+        let emsdk = env::var("EMSDK")
+            .expect("TARGET is emscripten but EMSDK is not set; needed for bindgen sysroot");
+        let sysroot = Path::new(&emsdk).join("upstream/emscripten/cache/sysroot");
+        if !sysroot.is_dir() {
+            panic!(
+                "TARGET is emscripten but EMSDK sysroot is missing: {}",
+                sysroot.display()
+            );
         }
+        builder = builder.clang_arg(format!("--sysroot={}", sysroot.display()));
     }
 
     let c_bindings = builder
@@ -85,13 +87,14 @@ fn build() -> bool {
         dst.generator("Ninja");
     }
 
+    dst.define("BUILD_CXX_EXE", "OFF");
+    dst.define("BUILD_EXAMPLES", "OFF");
+
     // `cmake` crate default C++ flags inject `-fno-exceptions` for this target,
     // but HiGHS requires C++ exceptions. Use explicit flags for emscripten.
     if emscripten {
         dst.no_default_flags(true);
         dst.cxxflag("-fexceptions");
-        dst.define("BUILD_CXX_EXE", "OFF");
-        dst.define("BUILD_EXAMPLES", "OFF");
     }
 
     // Avoid using downstream project's profile setting for HiGHS build.

--- a/build.rs
+++ b/build.rs
@@ -36,11 +36,24 @@ fn generate_bindings<'a>(
     // The bindgen::Builder is the main entry point
     // to bindgen, and lets you build up options for
     // the resulting bindings.
-    let builder = include_paths
+    let mut builder = include_paths
         .into_iter()
         .fold(bindgen::Builder::default(), |builder, path| {
             builder.clang_arg(format!("-I{}", path.to_string_lossy()))
         });
+    if env::var("TARGET")
+        .map(|target| target.contains("emscripten"))
+        .unwrap_or(false)
+    {
+        println!("cargo:rerun-if-env-changed=EMSDK");
+        if let Ok(emsdk) = env::var("EMSDK") {
+            let sysroot = Path::new(&emsdk).join("upstream/emscripten/cache/sysroot");
+            if sysroot.is_dir() {
+                builder = builder.clang_arg(format!("--sysroot={}", sysroot.display()));
+            }
+        }
+    }
+
     let c_bindings = builder
         // The input header we would like to generate bindings for.
         // This is a trivial wrapper header so that the HiGHS headers
@@ -64,10 +77,21 @@ fn generate_bindings<'a>(
 #[cfg(feature = "build")]
 fn build() -> bool {
     use cmake::Config;
+    let target = env::var("TARGET").unwrap();
+    let emscripten = target.contains("emscripten");
     let mut dst = Config::new("HiGHS");
 
     if cfg!(feature = "ninja") {
         dst.generator("Ninja");
+    }
+
+    // `cmake` crate default C++ flags inject `-fno-exceptions` for this target,
+    // but HiGHS requires C++ exceptions. Use explicit flags for emscripten.
+    if emscripten {
+        dst.no_default_flags(true);
+        dst.cxxflag("-fexceptions");
+        dst.define("BUILD_CXX_EXE", "OFF");
+        dst.define("BUILD_EXAMPLES", "OFF");
     }
 
     // Avoid using downstream project's profile setting for HiGHS build.
@@ -94,11 +118,10 @@ fn build() -> bool {
         println!("cargo:rustc-link-lib=z");
     }
 
-    let target = env::var("TARGET").unwrap();
     let apple = target.contains("apple");
     let linux = target.contains("linux");
     let mingw = target.contains("pc-windows-gnu");
-    if apple {
+    if apple || emscripten {
         println!("cargo:rustc-link-lib=c++");
     } else if linux || mingw {
         println!("cargo:rustc-link-lib=stdc++");

--- a/build.rs
+++ b/build.rs
@@ -33,28 +33,32 @@ fn generate_bindings<'a>(
     exclude_dir: Option<PathBuf>,
     include_paths: impl IntoIterator<Item = &'a Path>,
 ) {
+    let target = env::var("TARGET").unwrap();
+    let emscripten = target.contains("emscripten");
+
     // The bindgen::Builder is the main entry point
     // to bindgen, and lets you build up options for
     // the resulting bindings.
-    let mut builder = include_paths
+    let builder = include_paths
         .into_iter()
         .fold(bindgen::Builder::default(), |builder, path| {
             builder.clang_arg(format!("-I{}", path.to_string_lossy()))
         });
 
-    if env::var("TARGET").unwrap().contains("emscripten") {
-        println!("cargo:rerun-if-env-changed=EMSDK");
-        let emsdk = env::var("EMSDK")
-            .expect("TARGET is emscripten but EMSDK is not set; needed for bindgen sysroot");
-        let sysroot = Path::new(&emsdk).join("upstream/emscripten/cache/sysroot");
-        if !sysroot.is_dir() {
-            panic!(
-                "TARGET is emscripten but EMSDK sysroot is missing: {}",
-                sysroot.display()
-            );
-        }
-        builder = builder.clang_arg(format!("--sysroot={}", sysroot.display()));
-    }
+    // Work around bindgen/libclang missing `Highs_*` function declarations for
+    // `wasm32-unknown-emscripten` when parsing with emscripten's sysroot.
+    // The C API signatures are target-independent, so parse as host and skip
+    // cross-target layout assertions.
+    let builder = if emscripten {
+        println!("cargo:rerun-if-env-changed=HOST");
+        let host = env::var("HOST")
+            .expect("TARGET is emscripten but HOST is not set; needed for bindgen host parsing");
+        builder
+            .clang_arg(format!("--target={host}"))
+            .layout_tests(false)
+    } else {
+        builder
+    };
 
     let c_bindings = builder
         // The input header we would like to generate bindings for.

